### PR TITLE
Refactor Training Loop into Components

### PR DESCRIPTION
**Pull request by daveey**
Refactor Training Loop into Components

Now working.

| Component | Config(s) | Notes |
  |-----------|-----------|-------|
  | Heartbeat | HeartbeatConfig | Emits heartbeat pings. |
  | Scheduler | SchedulerConfig, HyperparameterSchedulerConfig | Drives LR/PPO schedules. |
  | GradientReporter | GradientReporterConfig | Logs gradient stats. |
  | Monitor | — | Keeps system/memory monitors running. |
  | StatsReporter | StatsReporterConfig, StatsReporterState | Ships metrics to wandb/stats; NoOpStatsReporter available. |
  | Uploader | UploaderConfig | Uploads policy artifacts. |
  | Checkpointer | CheckpointerConfig | Saves policies and tracks latest URI. |
  | ContextCheckpointer | ContextCheckpointerConfig | Persists trainer optimizer/timing state. |
  | TorchProfiler | TorchProfilerConfig | Runs periodic torch profiling sessions. |
  | WandbAborter | WandbAborterConfig | Stops training when wandb sets abort tag. |
  | WandbLogger | — | Logs epoch metrics and losses to wandb. |
  | Evaluator / NoOpEvaluator | EvaluatorConfig | Runs or skips policy evaluations. |

[Asana Task](https://app.asana.com/1/1209016784099267/project/1210348820405981/task/1211407068112464)

### CODE REVIEW

**Comment by daveey** [comment]
* **#2550** <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/2550?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Metta-AI/metta/2550?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: agent/src/metta/agent/policies/fast.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+    def reset_memory(self):
+        self.lstm.reset_memory()
+
+    def clip_weights(self) -> None:
+        for module in self._iter_component_modules():
+            if hasattr(module, "clip_weights"):
+                module.clip_weights()
+
+    def l2_init_loss(self) -> torch.Tensor:
```
Comment (Al-does) [inline agent/src/metta/agent/policies/fast.py:192]: How did these make it back?

File: agent/src/metta/agent/policies/fast.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    def compute_weight_metrics(self, delta: float = 0.01) -> List[dict]:
+        metrics: List[dict] = []
+        for module in self._components_with_metrics:
+            if module is not None and hasattr(module, "compute_weight_metrics"):
+                result = module.compute_weight_metrics(delta)
+                if result:
+                    metrics.append(result)
+        return metrics
+
+    def _iter_component_modules(self) -> List[nn.Module]:
```
Comment (Al-does) [inline agent/src/metta/agent/policies/fast.py:208]: Is this a new requirement of policies or just for this specific policy?

File: agent/src/metta/agent/policies/fast.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        td_modules = [
+            getattr(self.actor_1, "module", None),
+            getattr(self.critic_1, "module", None),
+            getattr(self.value_head, "module", None),
+        ]
+        modules.extend(td_modules)
+
+        return [module for module in modules if module is not None]
+
+    def _convert_action_to_logit_index(self, flattened_action: torch.Tensor) -> torch.Tensor:
```
Comment (Al-does) [inline agent/src/metta/agent/policies/fast.py:228]: These were moved to the actor layer. Do we prefer them to be back at the policy level?

**Review by Al-does — APPROVED** [review]
